pkg/object/mqttproxy/client.go

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mqttproxy
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/eclipse/paho.mqtt.golang/packets"
+	"github.com/megaease/easegress/pkg/logger"
+)
+
+const (
+	// Connected is MQTT client status of Connected
+	Connected = 1
+	// Disconnected is MQTT client status of Disconnected
+	Disconnected = 2
+
+	// Qos0 for "At most once"
+	Qos0 byte = 0
+	// Qos1 for "At least once
+	Qos1 byte = 1
+	// Qos2 for "Exactly once"
+	Qos2 byte = 2
+)
+
+type (
+	// ClientInfo is basic infomation for client
+	ClientInfo struct {
+		cid       string
+		username  string
+		password  string
+		keepalive uint16
+		will      *packets.PublishPacket
+	}
+
+	// Client represents a MQTT client connection in Broker
+	Client struct {
+		sync.Mutex
+
+		broker  *Broker
+		session *Session
+		conn    net.Conn
+
+		info   ClientInfo
+		status int
+		done   chan struct{}
+	}
+)
+
+func newClient(connect *packets.ConnectPacket, broker *Broker, conn net.Conn) *Client {
+	var will *packets.PublishPacket
+	if connect.WillFlag {
+		will = packets.NewControlPacket(packets.Publish).(*packets.PublishPacket)
+		will.Qos = connect.WillQos
+		will.TopicName = connect.WillTopic
+		will.Retain = connect.WillRetain
+		will.Payload = connect.WillMessage
+		will.Dup = connect.Dup
+	}
+
+	info := ClientInfo{
+		cid:       connect.ClientIdentifier,
+		username:  connect.Username,
+		password:  string(connect.Password),
+		keepalive: connect.Keepalive,
+		will:      will,
+	}
+	client := &Client{
+		broker: broker,
+		conn:   conn,
+		info:   info,
+		status: Connected,
+		done:   make(chan struct{}),
+	}
+	return client
+}
+
+func (c *Client) readLoop() {
+	defer c.close()
```

Comment:
At here, before disconnecting the network connection, you should handle `Will` packet. The implementation missed processing of `Will` flag, it isn't fulfilling the protocol requirement[1].

I think you should be strict compliance with protocol, please carefully checkout the protocol document.

[1] 3.1.2.5 in http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc398718030

Reply:
Thanks a lot for pointing out my huge mistake! I will fix that now!